// @selfxyz/agent-sdk/dist/agentCard.js
"use strict";
// SPDX-FileCopyrightText: 2025-2026 Social Connect Labs, Inc.
// SPDX-License-Identifier: BUSL-1.1
// NOTE: Converts to Apache-2.0 on 2029-06-11 per LICENSE.
Object.defineProperty(exports, "__esModule", { value: true });
exports.PROVIDER_LABELS = void 0;
exports.getProviderLabel = getProviderLabel;
exports.getStrengthColor = getStrengthColor;
exports.buildAgentCard = buildAgentCard;
exports.generateRegistrationJSON = generateRegistrationJSON;
// ─── Provider Scoring ────────────────────────────────────────────────────────
exports.PROVIDER_LABELS = {
    100: "passport",
    80: "kyc",
    60: "govt_id",
    40: "liveness",
};
/**
 * Map a numeric verification strength to a human-readable provider label.
 *
 * @param strength - Verification strength score (0-100) from the proof provider contract.
 * @returns A label such as "passport", "kyc", "govt_id", "liveness", or "unknown".
 */
function getProviderLabel(strength) {
    if (strength >= 100)
        return exports.PROVIDER_LABELS[100];
    if (strength >= 80)
        return exports.PROVIDER_LABELS[80];
    if (strength >= 60)
        return exports.PROVIDER_LABELS[60];
    if (strength >= 40)
        return exports.PROVIDER_LABELS[40];
    return "unknown";
}
/**
 * Map a numeric verification strength to a UI-friendly color tier.
 *
 * @param strength - Verification strength score (0-100).
 * @returns A color string: "green" (>=80), "blue" (>=60), "amber" (>=40), or "gray".
 */
function getStrengthColor(strength) {
    if (strength >= 80)
        return "green";
    if (strength >= 60)
        return "blue";
    if (strength >= 40)
        return "amber";
    return "gray";
}
// ─── Helpers ─────────────────────────────────────────────────────────────────
/**
 * Slugify a string into a URL/ID-safe lowercase-hyphenated form.
 * Used to auto-generate skill `id` values when not explicitly provided.
 */
function slugify(s) {
    return s
        .toLowerCase()
        .replace(/[^a-z0-9]+/g, "-")
        .replace(/^-|-$/g, "");
}
// ─── Build Card from On-Chain Data ───────────────────────────────────────────
/**
 * Build a complete ERC-8004 agent document by reading on-chain registry and provider data.
 *
 * Fetches the provider name, verification strength, and credential disclosures from
 * the contracts and merges them with the caller-supplied display fields.
 *
 * @param agentId - On-chain agent token ID.
 * @param registry - Ethers contract instance for the SelfAgentRegistry.
 * @param provider - Ethers contract instance for the SelfHumanProofProvider.
 * @param userFields - Display and service fields supplied by the caller.
 * @returns A fully populated {@link ERC8004AgentDocument}.
 */
async function buildAgentCard(agentId, registry, provider, userFields) {
    const [providerName, verificationStrength, credentials, registryAddress, chainId, proofProviderAddress,] = await Promise.all([
        provider.providerName(),
        provider.verificationStrength(),
        registry.getAgentCredentials(BigInt(agentId)).catch(() => null),
        registry.getAddress(),
        (registry.runner?.provider
            ?.getNetwork()
            .then((n) => Number(n.chainId)) ?? Promise.resolve(0)).catch(() => 0),
        provider.getAddress(),
    ]);
    const proofType = getProviderLabel(Number(verificationStrength));
    const strength = Number(verificationStrength);
    const trustModel = {
        proofType,
        sybilResistant: true,
        ofacScreened: false,
        minimumAgeVerified: 0,
    };
    let cardCredentials;
    if (credentials) {
        const olderThan = Number(credentials.olderThan ?? 0);
        const ofac = credentials.ofac ?? [false, false, false];
        const ofacScreened = ofac[0] === true;
        trustModel.ofacScreened = ofacScreened;
        trustModel.minimumAgeVerified = olderThan;
        cardCredentials = {
            nationality: credentials.nationality || undefined,
            issuingState: credentials.issuingState || undefined,
            olderThan: olderThan || undefined,
            ofacClean: ofacScreened || undefined,
            hasName: (credentials.name?.length ?? 0) > 0 ? true : undefined,
            hasDateOfBirth: credentials.dateOfBirth ? true : undefined,
            hasGender: credentials.gender ? true : undefined,
            documentExpiry: credentials.expiryDate || undefined,
        };
    }
    return {
        type: "https://eips.ethereum.org/EIPS/eip-8004#registration-v1",
        name: userFields.name,
        description: userFields.description ?? "",
        image: userFields.image ?? "",
        services: userFields.services ??
            (userFields.url
                ? [{ name: "A2A", endpoint: userFields.url, version: "1.0" }]
                : []),
        url: userFields.url,
        version: userFields.version,
        provider: userFields.agentProvider,
        capabilities: userFields.capabilities,
        securitySchemes: userFields.securitySchemes,
        security: userFields.security,
        supportedInterfaces: userFields.supportedInterfaces,
        iconUrl: userFields.iconUrl,
        documentationUrl: userFields.documentationUrl,
        signatures: userFields.signatures,
        extensions: userFields.extensions,
        skills: userFields.skills,
        selfProtocol: {
            agentId,
            registry: registryAddress,
            chainId,
            proofProvider: proofProviderAddress,
            providerName,
            verificationStrength: strength,
            trustModel,
            credentials: cardCredentials,
        },
    };
}
/**
 * Build an ERC-8004 registration JSON document synchronously from plain options.
 *
 * Unlike {@link buildAgentCard}, this does not read from on-chain contracts.
 * When `options.a2a` is provided, the returned document is also a valid A2A Agent Card.
 * Skills without an explicit `id` will have one auto-generated from the skill name.
 *
 * @param options - Document fields and optional A2A configuration.
 * @returns A populated {@link ERC8004AgentDocument}.
 */
function generateRegistrationJSON(options) {
    const a2a = options.a2a;
    // Auto-generate skill IDs if missing
    const skills = a2a?.skills?.map((skill) => ({
        ...skill,
        id: skill.id || slugify(skill.name),
    }));
    // Auto-generate supportedInterfaces from url if not explicitly provided
    const supportedInterfaces = a2a
        ? a2a.supportedInterfaces ?? [
            {
                url: a2a.url,
                protocolBinding: "JSONRPC",
                protocolVersion: "0.3.0",
            },
        ]
        : undefined;
    // Ensure services array contains an A2A entry when a2a is provided
    let services = options.services;
    if (a2a && !services.some((s) => s.name === "A2A")) {
        services = [
            ...services,
            { name: "A2A", endpoint: a2a.url, version: a2a.version },
        ];
    }
    return {
        type: "https://eips.ethereum.org/EIPS/eip-8004#registration-v1",
        name: options.name,
        description: options.description,
        image: options.image,
        services,
        active: options.active,
        registrations: options.registrations,
        supportedTrust: options.supportedTrust,
        ...(a2a && {
            version: a2a.version,
            url: a2a.url,
            provider: a2a.provider,
            capabilities: a2a.capabilities ?? {
                streaming: false,
                pushNotifications: false,
                stateTransitionHistory: false,
                extendedAgentCard: false,
            },
            // Include A2A v0.3.0 security scheme map
            ...(a2a.securitySchemes && {
                securitySchemes: a2a.securitySchemes,
            }),
            // Include security requirements
            ...(a2a.security && { security: a2a.security }),
            defaultInputModes: a2a.defaultInputModes,
            defaultOutputModes: a2a.defaultOutputModes,
            supportedInterfaces,
            iconUrl: a2a.iconUrl ?? options.image,
            documentationUrl: a2a.documentationUrl,
            signatures: a2a.signatures,
            extensions: a2a.extensions,
            skills,
        }),
    };
}
