// @selfxyz/agent-sdk/dist/Ed25519Agent.js
"use strict";
// SPDX-FileCopyrightText: 2025-2026 Social Connect Labs, Inc.
// SPDX-License-Identifier: BUSL-1.1
// NOTE: Converts to Apache-2.0 on 2029-06-11 per LICENSE.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ed25519Agent = void 0;
const ed = __importStar(require("@noble/ed25519"));
const node_crypto_1 = require("node:crypto");
const ethers_1 = require("ethers");
const constants_1 = require("./constants");
const signing_1 = require("./signing");
const contract_types_1 = require("./contract-types");
// @noble/ed25519 v3 requires a SHA-512 implementation for sync methods (getPublicKey, sign, verify).
// Use Node's built-in crypto module.
if (!ed.hashes.sha512) {
    ed.hashes.sha512 = (message) => new Uint8Array((0, node_crypto_1.createHash)("sha512").update(message).digest());
}
/**
 * Agent-side SDK for Self Agent ID using Ed25519 key pairs.
 *
 * The agent's on-chain identity is its raw 32-byte Ed25519 public key:
 *   agentKey = "0x" + hex(publicKey)
 *
 * For off-chain authentication, the agent signs each request with Ed25519.
 * Services verify the signature using the public key and check on-chain status.
 *
 * Usage:
 * ```ts
 * const agent = new Ed25519Agent({ privateKey: "0x..." });
 *
 * const registered = await agent.isRegistered();
 * const response = await agent.fetch("https://api.example.com/data");
 * ```
 */
class Ed25519Agent {
    constructor(config) {
        // Parse private key (strip 0x prefix if present)
        const privHex = config.privateKey.startsWith("0x")
            ? config.privateKey.slice(2)
            : config.privateKey;
        if (privHex.length !== 64) {
            throw new Error("Ed25519 private key must be 32 bytes (64 hex characters)");
        }
        this.privateKeyBytes = hexToBytes(privHex);
        this.publicKeyBytes = ed.getPublicKey(this.privateKeyBytes);
        const net = constants_1.NETWORKS[config.network ?? constants_1.DEFAULT_NETWORK];
        const provider = new ethers_1.ethers.JsonRpcProvider(config.rpcUrl ?? net.rpcUrl);
        this.registry = (0, contract_types_1.typedRegistry)(config.registryAddress ?? net.registryAddress, provider);
        // Agent key = raw 32-byte public key as 0x-prefixed hex (already 32 bytes)
        this._agentKey = "0x" + bytesToHex(this.publicKeyBytes);
        // Derive a deterministic Ethereum-style address from pubkey:
        // address = address(uint160(uint256(keccak256(pubkey))))
        this._address = Ed25519Agent.deriveAddress(this.publicKeyBytes);
    }
    /** The agent's on-chain key (bytes32) — raw Ed25519 public key */
    get agentKey() {
        return this._agentKey;
    }
    /**
     * A deterministic Ethereum-style address derived from keccak256(pubkey).
     * This matches the on-chain Ed25519Verifier.deriveAddress() function.
     */
    get address() {
        return this._address;
    }
    /** Check if this agent is registered and verified on-chain */
    async isRegistered() {
        return this.registry.isVerifiedAgent(this._agentKey);
    }
    /** Get full agent info from the registry */
    async getInfo() {
        const agentId = await this.registry.getAgentId(this._agentKey);
        if (agentId === 0n) {
            return {
                address: this._address,
                agentKey: this._agentKey,
                agentId: 0n,
                isVerified: false,
                nullifier: 0n,
                agentCount: 0n,
            };
        }
        const [isVerified, nullifier] = await Promise.all([
            this.registry.hasHumanProof(agentId),
            this.registry.getHumanNullifier(agentId),
        ]);
        const agentCount = await this.registry.getAgentCountForHuman(nullifier);
        return {
            address: this._address,
            agentKey: this._agentKey,
            agentId,
            isVerified,
            nullifier,
            agentCount,
        };
    }
    /**
     * Generate authentication headers for a request.
     *
     * Signature covers: keccak256(timestamp + method + canonicalPathAndQuery + bodyHash)
     * Signed with Ed25519 instead of ECDSA.
     */
    async signRequest(method, url, body) {
        const timestamp = Date.now().toString();
        const message = (0, signing_1.computeSigningMessage)(timestamp, method, url, body);
        // Sign the raw 32-byte keccak256 hash with Ed25519
        const msgBytes = ethers_1.ethers.getBytes(message);
        const sigBytes = await ed.signAsync(msgBytes, this.privateKeyBytes);
        return {
            [constants_1.HEADERS.KEY]: this._agentKey,
            [constants_1.HEADERS.KEYTYPE]: "ed25519",
            [constants_1.HEADERS.SIGNATURE]: "0x" + bytesToHex(sigBytes),
            [constants_1.HEADERS.TIMESTAMP]: timestamp,
        };
    }
    /**
     * Wrapper around global fetch that automatically adds agent signature headers.
     */
    async fetch(url, options = {}) {
        const method = (options.method || "GET").toUpperCase();
        const body = typeof options.body === "string" ? options.body : undefined;
        const headers = await this.signRequest(method, url, body);
        return globalThis.fetch(url, {
            ...options,
            method,
            headers: {
                ...Object.fromEntries(Object.entries(options.headers || {})),
                ...headers,
            },
        });
    }
    /**
     * Derive a deterministic Ethereum-style address from an Ed25519 public key.
     *
     * Matches the on-chain Ed25519Verifier.deriveAddress():
     *   address(uint160(uint256(keccak256(pubkey))))
     */
    static deriveAddress(pubkey) {
        const pubkeyBytes = typeof pubkey === "string" ? ethers_1.ethers.getBytes(pubkey) : pubkey;
        const hash = ethers_1.ethers.keccak256(pubkeyBytes);
        // Take the last 20 bytes of the 32-byte keccak256 hash
        return ethers_1.ethers.getAddress("0x" + hash.slice(-40));
    }
}
exports.Ed25519Agent = Ed25519Agent;
// ---------------------------------------------------------------------------
// Hex utility helpers (avoids Buffer dependency for browser compatibility)
// ---------------------------------------------------------------------------
function hexToBytes(hex) {
    const bytes = new Uint8Array(hex.length / 2);
    for (let i = 0; i < bytes.length; i++) {
        bytes[i] = parseInt(hex.slice(i * 2, i * 2 + 2), 16);
    }
    return bytes;
}
function bytesToHex(bytes) {
    let hex = "";
    for (let i = 0; i < bytes.length; i++) {
        hex += bytes[i].toString(16).padStart(2, "0");
    }
    return hex;
}
