// @selfxyz/agent-sdk/dist/Ed25519Agent.test.js
"use strict";
// SPDX-FileCopyrightText: 2025-2026 Social Connect Labs, Inc.
// SPDX-License-Identifier: BUSL-1.1
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_test_1 = require("node:test");
const strict_1 = __importDefault(require("node:assert/strict"));
const ed = __importStar(require("@noble/ed25519"));
const ethers_1 = require("ethers");
const Ed25519Agent_1 = require("./Ed25519Agent");
const signing_1 = require("./signing");
const constants_1 = require("./constants");
const SelfAgentVerifier_1 = require("./SelfAgentVerifier");
// A deterministic test private key (32 bytes hex)
const TEST_PRIVATE_KEY = "0x4cdb08e75df7b2cc tried9c3e5f8dc91a1b2c3d4e5f6a7b8c9d0e1f2a3b4c5d6e7f8";
// Use a simpler deterministic key
const TEST_KEY_HEX = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa";
(0, node_test_1.describe)("Ed25519Agent", () => {
    (0, node_test_1.it)("should construct from hex private key without 0x prefix", () => {
        const agent = new Ed25519Agent_1.Ed25519Agent({
            privateKey: TEST_KEY_HEX,
            rpcUrl: "http://localhost:1", // won't be used
            registryAddress: ethers_1.ethers.ZeroAddress,
        });
        strict_1.default.ok(agent.agentKey.startsWith("0x"));
        strict_1.default.equal(agent.agentKey.length, 66); // 0x + 64 hex chars = 32 bytes
    });
    (0, node_test_1.it)("should construct from hex private key with 0x prefix", () => {
        const agent = new Ed25519Agent_1.Ed25519Agent({
            privateKey: "0x" + TEST_KEY_HEX,
            rpcUrl: "http://localhost:1",
            registryAddress: ethers_1.ethers.ZeroAddress,
        });
        strict_1.default.ok(agent.agentKey.startsWith("0x"));
        strict_1.default.equal(agent.agentKey.length, 66);
    });
    (0, node_test_1.it)("should derive correct public key", () => {
        const privBytes = hexToBytes(TEST_KEY_HEX);
        const expectedPubkey = ed.getPublicKey(privBytes);
        const expectedKey = "0x" + bytesToHex(expectedPubkey);
        const agent = new Ed25519Agent_1.Ed25519Agent({
            privateKey: TEST_KEY_HEX,
            rpcUrl: "http://localhost:1",
            registryAddress: ethers_1.ethers.ZeroAddress,
        });
        strict_1.default.equal(agent.agentKey, expectedKey);
    });
    (0, node_test_1.it)("should derive deterministic address from keccak256(pubkey)", () => {
        const privBytes = hexToBytes(TEST_KEY_HEX);
        const pubkeyBytes = ed.getPublicKey(privBytes);
        const hash = ethers_1.ethers.keccak256(pubkeyBytes);
        const expectedAddress = ethers_1.ethers.getAddress("0x" + hash.slice(-40));
        const agent = new Ed25519Agent_1.Ed25519Agent({
            privateKey: TEST_KEY_HEX,
            rpcUrl: "http://localhost:1",
            registryAddress: ethers_1.ethers.ZeroAddress,
        });
        strict_1.default.equal(agent.address, expectedAddress);
    });
    (0, node_test_1.it)("should produce valid Ed25519 signatures in signRequest", async () => {
        const agent = new Ed25519Agent_1.Ed25519Agent({
            privateKey: TEST_KEY_HEX,
            rpcUrl: "http://localhost:1",
            registryAddress: ethers_1.ethers.ZeroAddress,
        });
        const headers = await agent.signRequest("GET", "https://example.com/api");
        // Check required headers are present
        strict_1.default.ok(headers[constants_1.HEADERS.KEY], "should have key header");
        strict_1.default.ok(headers[constants_1.HEADERS.KEYTYPE], "should have keytype header");
        strict_1.default.ok(headers[constants_1.HEADERS.SIGNATURE], "should have signature header");
        strict_1.default.ok(headers[constants_1.HEADERS.TIMESTAMP], "should have timestamp header");
        // Check keytype is ed25519
        strict_1.default.equal(headers[constants_1.HEADERS.KEYTYPE], "ed25519");
        // Check key matches agentKey
        strict_1.default.equal(headers[constants_1.HEADERS.KEY], agent.agentKey);
        // Verify the signature is valid
        const message = (0, signing_1.computeSigningMessage)(headers[constants_1.HEADERS.TIMESTAMP], "GET", "https://example.com/api");
        const msgBytes = ethers_1.ethers.getBytes(message);
        const sigBytes = ethers_1.ethers.getBytes(headers[constants_1.HEADERS.SIGNATURE]);
        const pubkeyBytes = ethers_1.ethers.getBytes(agent.agentKey);
        const valid = await ed.verifyAsync(sigBytes, msgBytes, pubkeyBytes);
        strict_1.default.ok(valid, "Ed25519 signature should be valid");
    });
    (0, node_test_1.it)("should produce valid signatures for POST requests with body", async () => {
        const agent = new Ed25519Agent_1.Ed25519Agent({
            privateKey: TEST_KEY_HEX,
            rpcUrl: "http://localhost:1",
            registryAddress: ethers_1.ethers.ZeroAddress,
        });
        const body = JSON.stringify({ hello: "world" });
        const headers = await agent.signRequest("POST", "https://example.com/api", body);
        const message = (0, signing_1.computeSigningMessage)(headers[constants_1.HEADERS.TIMESTAMP], "POST", "https://example.com/api", body);
        const msgBytes = ethers_1.ethers.getBytes(message);
        const sigBytes = ethers_1.ethers.getBytes(headers[constants_1.HEADERS.SIGNATURE]);
        const pubkeyBytes = ethers_1.ethers.getBytes(agent.agentKey);
        const valid = await ed.verifyAsync(sigBytes, msgBytes, pubkeyBytes);
        strict_1.default.ok(valid, "Ed25519 signature should be valid for POST with body");
    });
    (0, node_test_1.it)("should reject invalid private key length", () => {
        strict_1.default.throws(() => new Ed25519Agent_1.Ed25519Agent({
            privateKey: "0xdead",
            rpcUrl: "http://localhost:1",
            registryAddress: ethers_1.ethers.ZeroAddress,
        }), /32 bytes/);
    });
    (0, node_test_1.it)("deriveAddress static method should match instance address", () => {
        const privBytes = hexToBytes(TEST_KEY_HEX);
        const pubkeyBytes = ed.getPublicKey(privBytes);
        const staticAddr = Ed25519Agent_1.Ed25519Agent.deriveAddress(pubkeyBytes);
        const agent = new Ed25519Agent_1.Ed25519Agent({
            privateKey: TEST_KEY_HEX,
            rpcUrl: "http://localhost:1",
            registryAddress: ethers_1.ethers.ZeroAddress,
        });
        strict_1.default.equal(staticAddr, agent.address);
    });
    (0, node_test_1.it)("deriveAddress should accept hex string input", () => {
        const privBytes = hexToBytes(TEST_KEY_HEX);
        const pubkeyBytes = ed.getPublicKey(privBytes);
        const pubHex = "0x" + bytesToHex(pubkeyBytes);
        const fromBytes = Ed25519Agent_1.Ed25519Agent.deriveAddress(pubkeyBytes);
        const fromHex = Ed25519Agent_1.Ed25519Agent.deriveAddress(pubHex);
        strict_1.default.equal(fromBytes, fromHex);
    });
});
(0, node_test_1.describe)("SelfAgentVerifier Ed25519 path", () => {
    (0, node_test_1.it)("should verify a valid Ed25519 signed request", async () => {
        const agent = new Ed25519Agent_1.Ed25519Agent({
            privateKey: TEST_KEY_HEX,
            rpcUrl: "http://localhost:1",
            registryAddress: ethers_1.ethers.ZeroAddress,
        });
        const headers = await agent.signRequest("GET", "/api/test");
        // We can't test on-chain checks without a real provider, but we can verify
        // the signature verification path runs correctly up to the on-chain check.
        // Create a verifier with replay protection disabled and a short maxAge.
        const verifier = new SelfAgentVerifier_1.SelfAgentVerifier({
            rpcUrl: "http://localhost:1",
            registryAddress: ethers_1.ethers.ZeroAddress,
            enableReplayProtection: false,
            requireSelfProvider: false,
        });
        // Directly test verify — it will fail at the on-chain check (RPC error),
        // but we can at least verify the Ed25519 signature validation path works
        // by checking it doesn't fail with "Invalid Ed25519 signature"
        try {
            await verifier.verify({
                signature: headers[constants_1.HEADERS.SIGNATURE],
                timestamp: headers[constants_1.HEADERS.TIMESTAMP],
                method: "GET",
                url: "/api/test",
                keytype: "ed25519",
                agentKey: headers[constants_1.HEADERS.KEY],
            });
        }
        catch (err) {
            // RPC errors are expected — but signature errors mean our code is wrong
            const msg = err instanceof Error ? err.message : String(err);
            strict_1.default.ok(!msg.includes("Invalid Ed25519 signature"), "Should not fail on Ed25519 signature validation: " + msg);
        }
    });
    (0, node_test_1.it)("should reject Ed25519 request without agentKey", async () => {
        const verifier = new SelfAgentVerifier_1.SelfAgentVerifier({
            rpcUrl: "http://localhost:1",
            registryAddress: ethers_1.ethers.ZeroAddress,
            enableReplayProtection: false,
        });
        const result = await verifier.verify({
            signature: "0x" + "00".repeat(64),
            timestamp: Date.now().toString(),
            method: "GET",
            url: "/test",
            keytype: "ed25519",
            // agentKey intentionally omitted
        });
        strict_1.default.equal(result.valid, false);
        strict_1.default.ok(result.error?.includes("Missing agent key"));
    });
    (0, node_test_1.it)("should reject Ed25519 request with wrong signature", async () => {
        const agent = new Ed25519Agent_1.Ed25519Agent({
            privateKey: TEST_KEY_HEX,
            rpcUrl: "http://localhost:1",
            registryAddress: ethers_1.ethers.ZeroAddress,
        });
        const verifier = new SelfAgentVerifier_1.SelfAgentVerifier({
            rpcUrl: "http://localhost:1",
            registryAddress: ethers_1.ethers.ZeroAddress,
            enableReplayProtection: false,
        });
        const result = await verifier.verify({
            signature: "0x" + "ab".repeat(64), // bogus signature
            timestamp: Date.now().toString(),
            method: "GET",
            url: "/test",
            keytype: "ed25519",
            agentKey: agent.agentKey,
        });
        strict_1.default.equal(result.valid, false);
        strict_1.default.ok(result.error?.includes("Invalid Ed25519 signature"));
    });
});
// ---------------------------------------------------------------------------
// Hex utility helpers (matching Ed25519Agent's internal helpers)
// ---------------------------------------------------------------------------
function hexToBytes(hex) {
    const bytes = new Uint8Array(hex.length / 2);
    for (let i = 0; i < bytes.length; i++) {
        bytes[i] = parseInt(hex.slice(i * 2, i * 2 + 2), 16);
    }
    return bytes;
}
function bytesToHex(bytes) {
    let hex = "";
    for (let i = 0; i < bytes.length; i++) {
        hex += bytes[i].toString(16).padStart(2, "0");
    }
    return hex;
}
