// @selfxyz/agent-sdk/dist/registration-flow.js
"use strict";
// SPDX-FileCopyrightText: 2025-2026 Social Connect Labs, Inc.
// SPDX-License-Identifier: BUSL-1.1
// NOTE: Converts to Apache-2.0 on 2029-06-11 per LICENSE.
Object.defineProperty(exports, "__esModule", { value: true });
exports.RegistrationError = exports.ExpiredSessionError = void 0;
exports.requestRegistration = requestRegistration;
exports.requestDeregistration = requestDeregistration;
exports.requestProofRefresh = requestProofRefresh;
exports.getAgentInfo = getAgentInfo;
exports.getAgentsForHuman = getAgentsForHuman;
// ── Constants ────────────────────────────────────────────────────────────────
const DEFAULT_API_BASE = "https://self-agent-id.vercel.app";
const DEFAULT_POLL_INTERVAL_MS = 5000;
const DEFAULT_TIMEOUT_MS = 30 * 60 * 1000; // 30 minutes
const CHAIN_IDS = {
    mainnet: 42220,
    testnet: 11142220,
};
// ── Errors ───────────────────────────────────────────────────────────────────
class ExpiredSessionError extends Error {
    constructor(message = "Registration session expired. Call requestRegistration() again to start a new session.") {
        super(message);
        this.name = "ExpiredSessionError";
    }
}
exports.ExpiredSessionError = ExpiredSessionError;
class RegistrationError extends Error {
    constructor(message) {
        super(message);
        this.name = "RegistrationError";
    }
}
exports.RegistrationError = RegistrationError;
// ── Internal Helpers ─────────────────────────────────────────────────────────
/**
 * Read the API base URL from the SELF_AGENT_API_BASE environment variable,
 * falling back to the hardcoded default.
 *
 * @returns The base URL string.
 */
function defaultApiBaseFromEnv() {
    if (typeof process !== "undefined" && process?.env?.SELF_AGENT_API_BASE) {
        return process.env.SELF_AGENT_API_BASE;
    }
    return DEFAULT_API_BASE;
}
/**
 * Resolve the final API base URL, stripping any trailing slashes.
 *
 * @param apiBase - Explicit override, or undefined to use the environment/default.
 * @returns A normalized base URL with no trailing slash.
 */
function resolveApiBase(apiBase) {
    return (apiBase ?? defaultApiBaseFromEnv()).replace(/\/+$/, "");
}
function chainIdForNetwork(network) {
    return CHAIN_IDS[network];
}
async function apiFetch(url, init) {
    const res = await globalThis.fetch(url, init);
    const body = await res.json();
    if (!res.ok) {
        const errorMsg = typeof body === "object" && body !== null && "error" in body
            ? body.error
            : `HTTP ${res.status}`;
        throw new RegistrationError(errorMsg);
    }
    return body;
}
function sleep(ms) {
    return new Promise((resolve) => setTimeout(resolve, ms));
}
// ── Core Functions ───────────────────────────────────────────────────────────
/**
 * Initiate agent registration through the Self Agent ID REST API.
 *
 * Returns a session object with polling and key export methods.
 */
async function requestRegistration(opts) {
    const base = resolveApiBase(opts.apiBase);
    const network = opts.network ?? "mainnet";
    const payload = {
        mode: opts.mode,
        network,
        disclosures: opts.disclosures,
        humanAddress: opts.humanAddress,
        agentName: opts.agentName,
        agentDescription: opts.agentDescription,
    };
    const data = await apiFetch(`${base}/api/agent/register`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(payload),
    });
    return buildRegistrationSession(data, base);
}
/**
 * Initiate agent deregistration through the Self Agent ID REST API.
 */
async function requestDeregistration(opts) {
    const base = resolveApiBase(opts.apiBase);
    const network = opts.network ?? "mainnet";
    const payload = {
        network,
        agentAddress: opts.agentAddress,
        disclosures: opts.disclosures,
    };
    const data = await apiFetch(`${base}/api/agent/deregister`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(payload),
    });
    return buildDeregistrationSession(data, base);
}
/**
 * Initiate a proof refresh for an existing agent through the Self Agent ID REST API.
 *
 * Returns a session object with a deep link for the human to scan in the Self app,
 * and a polling method to wait for the new proof to be recorded on-chain.
 */
async function requestProofRefresh(opts) {
    const base = resolveApiBase(opts.apiBase);
    const network = opts.network ?? "mainnet";
    const payload = {
        agentId: opts.agentId,
        network,
        disclosures: opts.disclosures,
    };
    const data = await apiFetch(`${base}/api/agent/refresh`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(payload),
    });
    return buildRefreshSession(data, base);
}
/**
 * Fetch agent info from the public query API.
 */
async function getAgentInfo(agentId, opts) {
    const base = resolveApiBase(opts?.apiBase);
    const network = opts?.network ?? "mainnet";
    const chainId = chainIdForNetwork(network);
    return apiFetch(`${base}/api/agent/info/${chainId}/${agentId}`);
}
/**
 * Fetch all agents for a human address from the public query API.
 */
async function getAgentsForHuman(address, opts) {
    const base = resolveApiBase(opts?.apiBase);
    const network = opts?.network ?? "mainnet";
    const chainId = chainIdForNetwork(network);
    return apiFetch(`${base}/api/agent/agents/${chainId}/${address}`);
}
// ── Session Builders ─────────────────────────────────────────────────────────
function buildRegistrationSession(data, apiBase) {
    let currentToken = data.sessionToken;
    return {
        sessionToken: data.sessionToken,
        stage: data.stage,
        deepLink: data.deepLink,
        agentAddress: data.agentAddress,
        expiresAt: data.expiresAt,
        timeRemainingMs: data.timeRemainingMs,
        humanInstructions: data.humanInstructions,
        async waitForCompletion(opts) {
            const timeout = opts?.timeoutMs ?? DEFAULT_TIMEOUT_MS;
            const interval = opts?.pollIntervalMs ?? DEFAULT_POLL_INTERVAL_MS;
            const deadline = Date.now() + timeout;
            while (Date.now() < deadline) {
                await sleep(interval);
                let status;
                try {
                    status = await apiFetch(`${apiBase}/api/agent/register/status?token=${encodeURIComponent(currentToken)}`);
                }
                catch (err) {
                    if (err instanceof RegistrationError &&
                        err.message.toLowerCase().includes("expired")) {
                        throw new ExpiredSessionError();
                    }
                    throw err;
                }
                // Update the rolling token
                currentToken = status.sessionToken;
                if (status.stage === "completed") {
                    return {
                        agentId: status.agentId,
                        agentAddress: status.agentAddress ?? data.agentAddress,
                        credentials: status.credentials ?? undefined,
                        txHash: status.txHash ?? undefined,
                    };
                }
                if (status.stage === "failed") {
                    throw new RegistrationError("Registration failed on-chain.");
                }
                if (status.stage === "expired") {
                    throw new ExpiredSessionError();
                }
            }
            throw new RegistrationError(`Registration did not complete within ${timeout}ms. The session may still be active — call waitForCompletion() again to resume polling.`);
        },
        async exportKey() {
            const status = await apiFetch(`${apiBase}/api/agent/register/export`, {
                method: "POST",
                headers: { "Content-Type": "application/json" },
                body: JSON.stringify({ token: currentToken }),
            });
            return status.privateKey;
        },
    };
}
function buildRefreshSession(data, apiBase) {
    let currentToken = data.sessionToken;
    return {
        sessionToken: data.sessionToken,
        stage: data.stage,
        deepLink: data.deepLink,
        expiresAt: data.expiresAt,
        timeRemainingMs: data.timeRemainingMs,
        humanInstructions: data.humanInstructions,
        async waitForCompletion(opts) {
            const timeout = opts?.timeoutMs ?? DEFAULT_TIMEOUT_MS;
            const interval = opts?.pollIntervalMs ?? DEFAULT_POLL_INTERVAL_MS;
            const deadline = Date.now() + timeout;
            while (Date.now() < deadline) {
                await sleep(interval);
                let status;
                try {
                    status = await apiFetch(`${apiBase}/api/agent/refresh/status?token=${encodeURIComponent(currentToken)}`);
                }
                catch (err) {
                    if (err instanceof RegistrationError &&
                        err.message.toLowerCase().includes("expired")) {
                        throw new ExpiredSessionError("Proof refresh session expired. Call requestProofRefresh() again to start a new session.");
                    }
                    throw err;
                }
                currentToken = status.sessionToken;
                if (status.stage === "completed") {
                    // The status response includes the new proof expiry timestamp
                    const expiresAtRaw = status.proofExpiresAt;
                    const proofExpiresAt = expiresAtRaw
                        ? new Date(expiresAtRaw)
                        : new Date(Date.now() + 365 * 24 * 60 * 60 * 1000); // fallback: 1 year from now
                    return { proofExpiresAt };
                }
                if (status.stage === "failed") {
                    throw new RegistrationError("Proof refresh failed on-chain.");
                }
                if (status.stage === "expired") {
                    throw new ExpiredSessionError("Proof refresh session expired. Call requestProofRefresh() again to start a new session.");
                }
            }
            throw new RegistrationError(`Proof refresh did not complete within ${timeout}ms. The session may still be active — call waitForCompletion() again to resume polling.`);
        },
    };
}
function buildDeregistrationSession(data, apiBase) {
    let currentToken = data.sessionToken;
    return {
        sessionToken: data.sessionToken,
        stage: data.stage,
        deepLink: data.deepLink,
        expiresAt: data.expiresAt,
        timeRemainingMs: data.timeRemainingMs,
        humanInstructions: data.humanInstructions,
        async waitForCompletion(opts) {
            const timeout = opts?.timeoutMs ?? DEFAULT_TIMEOUT_MS;
            const interval = opts?.pollIntervalMs ?? DEFAULT_POLL_INTERVAL_MS;
            const deadline = Date.now() + timeout;
            while (Date.now() < deadline) {
                await sleep(interval);
                let status;
                try {
                    status = await apiFetch(`${apiBase}/api/agent/deregister/status?token=${encodeURIComponent(currentToken)}`);
                }
                catch (err) {
                    if (err instanceof RegistrationError &&
                        err.message.toLowerCase().includes("expired")) {
                        throw new ExpiredSessionError("Deregistration session expired. Call requestDeregistration() again to start a new session.");
                    }
                    throw err;
                }
                currentToken = status.sessionToken;
                if (status.stage === "completed") {
                    return;
                }
                if (status.stage === "failed") {
                    throw new RegistrationError("Deregistration failed on-chain.");
                }
                if (status.stage === "expired") {
                    throw new ExpiredSessionError("Deregistration session expired. Call requestDeregistration() again to start a new session.");
                }
            }
            throw new RegistrationError(`Deregistration did not complete within ${timeout}ms. The session may still be active — call waitForCompletion() again to resume polling.`);
        },
    };
}
