// @selfxyz/agent-sdk/dist/SelfAgentVerifier.js
"use strict";
// SPDX-FileCopyrightText: 2025-2026 Social Connect Labs, Inc.
// SPDX-License-Identifier: BUSL-1.1
// NOTE: Converts to Apache-2.0 on 2029-06-11 per LICENSE.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SelfAgentVerifier = exports.VerifierBuilder = void 0;
exports.verifyAgent = verifyAgent;
const ed = __importStar(require("@noble/ed25519"));
const node_crypto_1 = require("node:crypto");
const ethers_1 = require("ethers");
// @noble/ed25519 v3 requires a SHA-512 implementation for sync methods.
if (!ed.hashes.sha512) {
    ed.hashes.sha512 = (message) => new Uint8Array((0, node_crypto_1.createHash)("sha512").update(message).digest());
}
const constants_1 = require("./constants");
const signing_1 = require("./signing");
const contract_types_1 = require("./contract-types");
class RateLimiter {
    constructor(config) {
        this.buckets = new Map();
        this.perMinute = config.perMinute ?? 0;
        this.perHour = config.perHour ?? 0;
    }
    /** Returns null if allowed, or { error, retryAfterMs } if rate limited. */
    check(agentAddress) {
        const now = Date.now();
        const key = agentAddress.toLowerCase();
        let bucket = this.buckets.get(key);
        if (!bucket) {
            bucket = { timestamps: [] };
            this.buckets.set(key, bucket);
        }
        // Prune timestamps older than 1 hour (longest window we care about)
        const oneHourAgo = now - 60 * 60 * 1000;
        bucket.timestamps = bucket.timestamps.filter((t) => t > oneHourAgo);
        // Check per-minute limit
        if (this.perMinute > 0) {
            const oneMinuteAgo = now - 60 * 1000;
            const recentMinute = bucket.timestamps.filter((t) => t > oneMinuteAgo);
            if (recentMinute.length >= this.perMinute) {
                const oldest = recentMinute[0];
                const retryAfterMs = oldest + 60 * 1000 - now;
                return {
                    error: `Rate limit exceeded (${this.perMinute}/min)`,
                    retryAfterMs: Math.max(1, retryAfterMs),
                };
            }
        }
        // Check per-hour limit
        if (this.perHour > 0) {
            if (bucket.timestamps.length >= this.perHour) {
                const oldest = bucket.timestamps[0];
                const retryAfterMs = oldest + 60 * 60 * 1000 - now;
                return {
                    error: `Rate limit exceeded (${this.perHour}/hr)`,
                    retryAfterMs: Math.max(1, retryAfterMs),
                };
            }
        }
        // Record this request
        bucket.timestamps.push(now);
        return null;
    }
}
// ---------------------------------------------------------------------------
// VerifierBuilder — chainable builder API
// ---------------------------------------------------------------------------
class VerifierBuilder {
    /** Set the network: "mainnet" or "testnet" */
    network(name) {
        this._network = name;
        return this;
    }
    /** Set a custom registry address */
    registry(addr) {
        this._registryAddress = addr;
        return this;
    }
    /** Set a custom RPC URL */
    rpc(url) {
        this._rpcUrl = url;
        return this;
    }
    /** Require the agent's human to be at least `n` years old */
    requireAge(n) {
        this._minimumAge = n;
        return this;
    }
    /** Require OFAC screening passed */
    requireOFAC() {
        this._requireOFACPassed = true;
        return this;
    }
    /** Require nationality in the given list */
    requireNationality(...codes) {
        this._allowedNationalities = codes;
        return this;
    }
    /** Require Self Protocol as proof provider (default: on) */
    requireSelfProvider() {
        this._requireSelfProvider = true;
        return this;
    }
    /** Max agents per human (default: 1) */
    sybilLimit(n) {
        this._maxAgentsPerHuman = n;
        return this;
    }
    /** Enable in-memory per-agent rate limiting */
    rateLimit(config) {
        this._rateLimitConfig = config;
        return this;
    }
    /** Enable replay protection (default: on) */
    replayProtection(enabled = true) {
        this._enableReplayProtection = enabled;
        return this;
    }
    /** Include ZK credentials in verification result */
    includeCredentials() {
        this._includeCredentials = true;
        return this;
    }
    /** Max signed timestamp age in milliseconds */
    maxAge(ms) {
        this._maxAgeMs = ms;
        return this;
    }
    /** On-chain cache TTL in milliseconds */
    cacheTtl(ms) {
        this._cacheTtlMs = ms;
        return this;
    }
    /** Build the SelfAgentVerifier instance */
    build() {
        // Auto-enable credentials if any credential requirement is set
        const needsCredentials = this._minimumAge != null ||
            this._requireOFACPassed ||
            (this._allowedNationalities && this._allowedNationalities.length > 0);
        return new SelfAgentVerifier({
            network: this._network,
            registryAddress: this._registryAddress,
            rpcUrl: this._rpcUrl,
            maxAgeMs: this._maxAgeMs,
            cacheTtlMs: this._cacheTtlMs,
            maxAgentsPerHuman: this._maxAgentsPerHuman,
            includeCredentials: needsCredentials || this._includeCredentials || undefined,
            requireSelfProvider: this._requireSelfProvider,
            enableReplayProtection: this._enableReplayProtection,
            minimumAge: this._minimumAge,
            requireOFACPassed: this._requireOFACPassed,
            allowedNationalities: this._allowedNationalities,
            rateLimitConfig: this._rateLimitConfig,
        });
    }
}
exports.VerifierBuilder = VerifierBuilder;
// ---------------------------------------------------------------------------
// SelfAgentVerifier
// ---------------------------------------------------------------------------
/**
 * Service-side verifier for Self Agent ID requests.
 *
 * Security chain:
 * 1. Recover signer address from ECDSA signature (cryptographic proof of key ownership)
 * 2. Derive agent key: zeroPadValue(recoveredAddress, 32)
 * 3. Check on-chain: isVerifiedAgent(agentKey) (proof that a human registered this address)
 * 4. Check proof provider: getProofProvider(agentId) matches selfProofProvider()
 * 5. Check timestamp freshness (replay protection)
 *
 * The signer address is RECOVERED from the signature, never trusted from headers.
 * This closes the off-chain verification gap — you can't claim to be an agent
 * without holding its private key.
 *
 * Usage:
 * ```ts
 * // Mainnet (default — no config needed)
 * const verifier = new SelfAgentVerifier();
 *
 * // Testnet
 * const verifier = new SelfAgentVerifier({ network: "testnet" });
 *
 * // Chainable builder
 * const verifier = SelfAgentVerifier.create()
 *   .network("testnet")
 *   .requireAge(18)
 *   .requireOFAC()
 *   .rateLimit({ perMinute: 10 })
 *   .build();
 *
 * // From config object
 * const verifier = SelfAgentVerifier.fromConfig({
 *   network: "testnet",
 *   requireAge: 18,
 *   requireOFAC: true,
 * });
 *
 * const result = await verifier.verify({
 *   signature: req.headers["x-self-agent-signature"],
 *   timestamp: req.headers["x-self-agent-timestamp"],
 *   method: req.method,
 *   url: req.originalUrl,
 *   body: req.body ? JSON.stringify(req.body) : undefined,
 * });
 *
 * if (result.valid) {
 *   console.log("Verified agent:", result.agentAddress);
 * }
 * ```
 */
class SelfAgentVerifier {
    constructor(config = {}) {
        this.cache = new Map();
        this.replayCache = new Map();
        this.selfProviderCache = null;
        const net = constants_1.NETWORKS[config.network ?? constants_1.DEFAULT_NETWORK];
        const provider = new ethers_1.ethers.JsonRpcProvider(config.rpcUrl ?? net.rpcUrl);
        this.registry = (0, contract_types_1.typedRegistry)(config.registryAddress ?? net.registryAddress, provider);
        this.maxAgeMs = config.maxAgeMs ?? constants_1.DEFAULT_MAX_AGE_MS;
        this.cacheTtlMs = config.cacheTtlMs ?? constants_1.DEFAULT_CACHE_TTL_MS;
        this.maxAgentsPerHuman = config.maxAgentsPerHuman ?? 1;
        this.includeCredentials = config.includeCredentials ?? false;
        this.requireSelfProvider = config.requireSelfProvider ?? true;
        this.enableReplayProtection = config.enableReplayProtection ?? true;
        this.replayCacheMaxEntries = config.replayCacheMaxEntries ?? 10000;
        this.minimumAge = config.minimumAge;
        this.requireOFACPassed = config.requireOFACPassed ?? false;
        this.allowedNationalities = config.allowedNationalities;
        this.rateLimiter = config.rateLimitConfig
            ? new RateLimiter(config.rateLimitConfig)
            : null;
    }
    /** Create a chainable builder for configuring a verifier */
    static create() {
        return new VerifierBuilder();
    }
    /** Create a verifier from a flat config object */
    static fromConfig(cfg) {
        // Auto-enable credentials if any credential requirement is set
        const needsCredentials = cfg.requireAge != null ||
            cfg.requireOFAC ||
            (cfg.requireNationality && cfg.requireNationality.length > 0);
        return new SelfAgentVerifier({
            network: cfg.network,
            registryAddress: cfg.registryAddress,
            rpcUrl: cfg.rpcUrl,
            maxAgeMs: cfg.maxAgeMs,
            cacheTtlMs: cfg.cacheTtlMs,
            maxAgentsPerHuman: cfg.sybilLimit,
            includeCredentials: needsCredentials || undefined,
            requireSelfProvider: cfg.requireSelfProvider,
            enableReplayProtection: cfg.replayProtection,
            minimumAge: cfg.requireAge,
            requireOFACPassed: cfg.requireOFAC,
            allowedNationalities: cfg.requireNationality,
            rateLimitConfig: cfg.rateLimit,
        });
    }
    /**
     * Verify a signed agent request.
     *
     * The agent's identity is derived from the signature itself — not from
     * any header that could be spoofed. This is the key security property.
     */
    async verify(params) {
        const { signature, timestamp, method, url, body, keytype } = params;
        const empty = {
            valid: false,
            agentAddress: ethers_1.ethers.ZeroAddress,
            agentKey: ethers_1.ethers.ZeroHash,
            agentId: 0n,
            agentCount: 0n,
            nullifier: 0n,
        };
        // 1. Check timestamp freshness (replay protection)
        const ts = parseInt(timestamp, 10);
        if (isNaN(ts) || Math.abs(Date.now() - ts) > this.maxAgeMs) {
            return { ...empty, error: "Timestamp expired or invalid" };
        }
        // 2. Reconstruct the signed message
        const canonicalUrl = (0, signing_1.canonicalizeSigningUrl)(url);
        const message = (0, signing_1.computeSigningMessage)(timestamp, method, canonicalUrl, body);
        let signerAddress;
        let agentKey;
        if (keytype === "ed25519") {
            // ── Ed25519 verification path ──
            if (!params.agentKey) {
                return { ...empty, error: "Missing agent key for Ed25519 verification" };
            }
            // Verify Ed25519 signature directly against the provided pubkey
            try {
                const msgBytes = ethers_1.ethers.getBytes(message); // 32-byte keccak256 hash
                const sigBytes = ethers_1.ethers.getBytes(signature); // 64-byte Ed25519 signature
                const pubkeyBytes = ethers_1.ethers.getBytes(params.agentKey); // 32-byte Ed25519 pubkey
                const valid = await ed.verifyAsync(sigBytes, msgBytes, pubkeyBytes);
                if (!valid) {
                    return { ...empty, error: "Invalid Ed25519 signature" };
                }
            }
            catch {
                return { ...empty, error: "Invalid Ed25519 signature" };
            }
            agentKey = params.agentKey;
            // Derive deterministic address from keccak256(pubkey) — matches Ed25519Verifier.deriveAddress on-chain
            const pubkeyBytes = ethers_1.ethers.getBytes(params.agentKey);
            const hash = ethers_1.ethers.keccak256(pubkeyBytes);
            signerAddress = ethers_1.ethers.getAddress("0x" + hash.slice(-40));
        }
        else {
            // ── secp256k1 ECDSA verification path (unchanged) ──
            // 3. Recover signer address from signature (cryptographic — can't be faked)
            try {
                signerAddress = ethers_1.ethers.verifyMessage(ethers_1.ethers.getBytes(message), signature);
            }
            catch {
                return { ...empty, error: "Invalid signature" };
            }
            // 5. Derive the on-chain agent key from the recovered address
            agentKey = ethers_1.ethers.zeroPadValue(signerAddress, 32);
        }
        // 4. Replay cache check (after signature validity to avoid cache poisoning)
        if (this.enableReplayProtection) {
            const replayError = this.checkAndRecordReplay(signature, message, ts);
            if (replayError) {
                return {
                    ...empty,
                    agentAddress: signerAddress,
                    agentKey,
                    error: replayError,
                };
            }
        }
        // 6. Check on-chain status (with cache)
        const { isVerified, isProofFresh, agentId, agentCount, nullifier, providerAddress, proofExpiresAtTimestamp, } = await this.checkOnChain(agentKey);
        // Compute expiry fields from on-chain timestamp
        const proofExpiresAt = proofExpiresAtTimestamp > 0n
            ? new Date(Number(proofExpiresAtTimestamp) * 1000)
            : undefined;
        const now = Math.floor(Date.now() / 1000);
        const daysUntilExpiry = proofExpiresAtTimestamp > 0n
            ? Math.floor((Number(proofExpiresAtTimestamp) - now) / 86400)
            : undefined;
        const isExpiringSoon = daysUntilExpiry !== undefined && daysUntilExpiry >= 0 && daysUntilExpiry <= 30;
        if (!isVerified) {
            return {
                valid: false,
                agentAddress: signerAddress,
                agentKey,
                agentId,
                agentCount,
                nullifier,
                proofExpiresAt,
                daysUntilExpiry,
                isExpiringSoon,
                error: "Agent not verified on-chain",
            };
        }
        // 6b. Check proof freshness (expired proofs should not pass verification)
        if (!isProofFresh) {
            return {
                valid: false,
                agentAddress: signerAddress,
                agentKey,
                agentId,
                agentCount,
                nullifier,
                proofExpiresAt,
                daysUntilExpiry,
                isExpiringSoon,
                error: "Agent's human proof has expired",
            };
        }
        // 7. Provider check: ensure agent was verified by Self Protocol
        if (this.requireSelfProvider && agentId > 0n) {
            let selfProvider;
            try {
                selfProvider = await this.getSelfProviderAddress();
            }
            catch {
                return {
                    valid: false,
                    agentAddress: signerAddress,
                    agentKey,
                    agentId,
                    agentCount,
                    nullifier,
                    proofExpiresAt,
                    daysUntilExpiry,
                    isExpiringSoon,
                    error: "Unable to verify proof provider — RPC error",
                };
            }
            if (providerAddress.toLowerCase() !== selfProvider.toLowerCase()) {
                return {
                    valid: false,
                    agentAddress: signerAddress,
                    agentKey,
                    agentId,
                    agentCount,
                    nullifier,
                    proofExpiresAt,
                    daysUntilExpiry,
                    isExpiringSoon,
                    error: "Agent was not verified by Self — proof provider mismatch",
                };
            }
        }
        // 8. Sybil resistance: reject if human has too many agents
        if (this.maxAgentsPerHuman > 0 &&
            agentCount > BigInt(this.maxAgentsPerHuman)) {
            return {
                valid: false,
                agentAddress: signerAddress,
                agentKey,
                agentId,
                agentCount,
                nullifier,
                proofExpiresAt,
                daysUntilExpiry,
                isExpiringSoon,
                error: `Human has ${agentCount} agents (max ${this.maxAgentsPerHuman})`,
            };
        }
        // 9. Fetch credentials if requested
        let credentials;
        if (this.includeCredentials && agentId > 0n) {
            credentials = await this.fetchCredentials(agentId);
        }
        // 10. Credential checks (post-verify — only if credentials were fetched)
        if (credentials) {
            if (this.minimumAge != null &&
                credentials.olderThan < BigInt(this.minimumAge)) {
                return {
                    valid: false,
                    agentAddress: signerAddress,
                    agentKey,
                    agentId,
                    agentCount,
                    nullifier,
                    credentials,
                    proofExpiresAt,
                    daysUntilExpiry,
                    isExpiringSoon,
                    error: `Agent's human does not meet minimum age (required: ${this.minimumAge}, got: ${credentials.olderThan})`,
                };
            }
            if (this.requireOFACPassed && !credentials.ofac?.[0]) {
                return {
                    valid: false,
                    agentAddress: signerAddress,
                    agentKey,
                    agentId,
                    agentCount,
                    nullifier,
                    credentials,
                    proofExpiresAt,
                    daysUntilExpiry,
                    isExpiringSoon,
                    error: "Agent's human did not pass OFAC screening",
                };
            }
            if (this.allowedNationalities && this.allowedNationalities.length > 0) {
                if (!this.allowedNationalities.includes(credentials.nationality)) {
                    return {
                        valid: false,
                        agentAddress: signerAddress,
                        agentKey,
                        agentId,
                        agentCount,
                        nullifier,
                        credentials,
                        proofExpiresAt,
                        daysUntilExpiry,
                        isExpiringSoon,
                        error: `Nationality "${credentials.nationality}" not in allowed list`,
                    };
                }
            }
        }
        // 11. Rate limiting (per-agent, in-memory sliding window)
        if (this.rateLimiter) {
            const limited = this.rateLimiter.check(signerAddress);
            if (limited) {
                return {
                    valid: false,
                    agentAddress: signerAddress,
                    agentKey,
                    agentId,
                    agentCount,
                    nullifier,
                    credentials,
                    proofExpiresAt,
                    daysUntilExpiry,
                    isExpiringSoon,
                    error: limited.error,
                    retryAfterMs: limited.retryAfterMs,
                };
            }
        }
        return {
            valid: true,
            agentAddress: signerAddress,
            agentKey,
            agentId,
            agentCount,
            nullifier,
            credentials,
            proofExpiresAt,
            daysUntilExpiry,
            isExpiringSoon,
        };
    }
    /**
     * Check on-chain agent status with caching.
     */
    async checkOnChain(agentKey) {
        const cached = this.cache.get(agentKey);
        if (cached && cached.expiresAt > Date.now()) {
            return {
                isVerified: cached.isVerified,
                isProofFresh: cached.isProofFresh,
                agentId: cached.agentId,
                agentCount: cached.agentCount,
                nullifier: cached.nullifier,
                providerAddress: cached.providerAddress,
                proofExpiresAtTimestamp: cached.proofExpiresAtTimestamp,
            };
        }
        const [isVerified, agentId] = await Promise.all([
            this.registry.isVerifiedAgent(agentKey),
            this.registry.getAgentId(agentKey),
        ]);
        // Fetch sybil data, provider address, proof freshness, and expiry if agent exists
        let agentCount = 0n;
        let nullifier = 0n;
        let providerAddress = "";
        let isProofFresh = false;
        let proofExpiresAtTimestamp = 0n;
        if (agentId > 0n) {
            const promises = [];
            // Always check proof freshness
            promises.push(this.registry.isProofFresh(agentId).then((fresh) => {
                isProofFresh = fresh;
            }));
            // Always fetch proof expiry timestamp
            promises.push(this.registry.proofExpiresAt(agentId).then((ts) => {
                proofExpiresAtTimestamp = ts;
            }));
            if (this.maxAgentsPerHuman > 0) {
                promises.push(this.registry.getHumanNullifier(agentId).then(async (n) => {
                    nullifier = n;
                    agentCount = await this.registry.getAgentCountForHuman(n);
                }));
            }
            if (this.requireSelfProvider) {
                promises.push(this.registry.getProofProvider(agentId).then((addr) => {
                    providerAddress = addr;
                }));
            }
            await Promise.all(promises);
        }
        this.cache.set(agentKey, {
            isVerified,
            isProofFresh,
            agentId,
            agentCount,
            nullifier,
            providerAddress,
            proofExpiresAtTimestamp,
            expiresAt: Date.now() + this.cacheTtlMs,
        });
        return {
            isVerified,
            isProofFresh,
            agentId,
            agentCount,
            nullifier,
            providerAddress,
            proofExpiresAtTimestamp,
        };
    }
    /**
     * Get Self Protocol's own proof provider address from the registry.
     * Cached separately since it rarely changes.
     */
    async getSelfProviderAddress() {
        if (this.selfProviderCache &&
            this.selfProviderCache.expiresAt > Date.now()) {
            return this.selfProviderCache.address;
        }
        // No try/catch — let RPC errors propagate to fail closed
        const address = await this.registry.selfProofProvider();
        this.selfProviderCache = {
            address,
            expiresAt: Date.now() + this.cacheTtlMs * 12, // Cache for longer (1 hour at default TTL)
        };
        return address;
    }
    /**
     * Fetch ZK-attested credentials for an agent.
     */
    async fetchCredentials(agentId) {
        try {
            const raw = await this.registry.getAgentCredentials(agentId);
            return {
                issuingState: raw.issuingState ?? "",
                name: raw.name ?? [],
                idNumber: raw.idNumber ?? "",
                nationality: raw.nationality ?? "",
                dateOfBirth: raw.dateOfBirth ?? "",
                gender: raw.gender ?? "",
                expiryDate: raw.expiryDate ?? "",
                olderThan: raw.olderThan ?? 0n,
                ofac: raw.ofac ?? [false, false, false],
            };
        }
        catch {
            return undefined;
        }
    }
    /** Clear the on-chain status cache */
    clearCache() {
        this.cache.clear();
        this.selfProviderCache = null;
        this.replayCache.clear();
    }
    checkAndRecordReplay(signature, message, ts) {
        const now = Date.now();
        this.pruneReplayCache(now);
        const key = `${signature.toLowerCase()}:${message.toLowerCase()}`;
        const existing = this.replayCache.get(key);
        if (existing && existing.expiresAt > now) {
            return "Replay detected";
        }
        const expiresAt = ts + this.maxAgeMs;
        this.replayCache.set(key, { expiresAt });
        return null;
    }
    pruneReplayCache(now) {
        for (const [key, entry] of this.replayCache) {
            if (entry.expiresAt <= now) {
                this.replayCache.delete(key);
            }
        }
        if (this.replayCache.size <= this.replayCacheMaxEntries)
            return;
        // Drop oldest-ish entries by earliest expiration first.
        const overflow = this.replayCache.size - this.replayCacheMaxEntries;
        const sorted = [...this.replayCache.entries()].sort((a, b) => a[1].expiresAt - b[1].expiresAt);
        for (let i = 0; i < overflow; i++) {
            this.replayCache.delete(sorted[i][0]);
        }
    }
    /**
     * Express/Connect middleware that verifies agent requests.
     *
     * Adds `req.agent` with `{ address, agentKey, agentId, agentCount, nullifier, credentials? }` on success.
     * Returns 401 on failure.
     *
     * Usage:
     * ```ts
     * app.use("/api", verifier.auth());
     * ```
     */
    auth() {
        return async (req, res, next) => {
            const signatureHeader = req.headers?.[constants_1.HEADERS.SIGNATURE];
            const timestampHeader = req.headers?.[constants_1.HEADERS.TIMESTAMP];
            const keytypeHeader = req.headers?.[constants_1.HEADERS.KEYTYPE];
            const keyHeader = req.headers?.[constants_1.HEADERS.KEY];
            const signature = typeof signatureHeader === "string" ? signatureHeader : undefined;
            const timestamp = typeof timestampHeader === "string" ? timestampHeader : undefined;
            const keytype = typeof keytypeHeader === "string" ? keytypeHeader : undefined;
            const agentKey = typeof keyHeader === "string" ? keyHeader : undefined;
            if (!signature || !timestamp) {
                res.status(401).json({ error: "Missing agent authentication headers" });
                return;
            }
            const method = typeof req.method === "string" ? req.method : "GET";
            const url = typeof req.originalUrl === "string"
                ? req.originalUrl
                : typeof req.url === "string"
                    ? req.url
                    : "/";
            const body = typeof req.rawBody === "string"
                ? req.rawBody
                : Buffer.isBuffer(req.rawBody)
                    ? req.rawBody.toString("utf8")
                    : typeof req.body === "string"
                        ? req.body
                        : req.body
                            ? JSON.stringify(req.body)
                            : undefined;
            const result = await this.verify({
                signature,
                timestamp,
                method,
                url,
                body,
                keytype,
                agentKey,
            });
            if (!result.valid) {
                const status = result.retryAfterMs ? 429 : 401;
                const body = { error: result.error };
                if (result.retryAfterMs) {
                    body.retryAfterMs = result.retryAfterMs;
                }
                res.status(status).json(body);
                return;
            }
            req.agent = {
                address: result.agentAddress,
                agentKey: result.agentKey,
                agentId: result.agentId,
                agentCount: result.agentCount,
                nullifier: result.nullifier,
                credentials: result.credentials,
            };
            next();
        };
    }
    /**
     * @deprecated Use `auth()` instead.
     */
    expressMiddleware() {
        return this.auth();
    }
}
exports.SelfAgentVerifier = SelfAgentVerifier;
// ---------------------------------------------------------------------------
// Standalone verifyAgent() — lightweight proof-expiry-aware check
// ---------------------------------------------------------------------------
/**
 * Build a re-authentication URL that the agent operator can visit to renew
 * their expired (or soon-to-expire) human proof.
 */
function buildReauthUrl(agentId, options) {
    const base = options.reauthBaseUrl ?? constants_1.REAUTH_BASE_URL;
    return `${base}/reauth?agentId=${agentId}&chainId=${options.chainId}&registry=${options.registryAddress}`;
}
/**
 * Standalone proof-expiry-aware agent verification.
 *
 * Unlike `SelfAgentVerifier.verify()` (which validates ECDSA request
 * signatures), this function performs a direct on-chain lookup to answer
 * the question: "does this agent currently hold a valid, non-expired
 * human proof?"
 *
 * It is the recommended entry point for ERC-8004 compliance checks when
 * you already trust the agent's identity (e.g., during off-chain enrollment
 * or administrative tooling).
 *
 * @param agentKey - The agent's bytes32 on-chain key (zero-padded address)
 * @param options  - `chainId` (used in reauth URL) and `registryAddress`
 * @param rpcUrl   - RPC endpoint to use (default: Celo mainnet)
 *
 * @returns A {@link VerifyResult} discriminated union:
 *   - `{ verified: true, agentId, expiresAt }` — active proof
 *   - `{ verified: false, reason: 'NOT_REGISTERED' }` — unknown key
 *   - `{ verified: false, reason: 'NO_HUMAN_PROOF' }` — key registered but no proof
 *   - `{ verified: false, reason: 'PROOF_EXPIRED', expiredAt, reauthUrl }` — proof lapsed
 *
 * @example
 * ```ts
 * import { verifyAgent, isProofExpiringSoon } from "@selfxyz/agent-sdk";
 *
 * const result = await verifyAgent(agentKey, { chainId: 42220, registryAddress: "0x..." });
 * if (!result.verified) {
 *   if (result.reason === "PROOF_EXPIRED") {
 *     console.warn("Re-auth at:", result.reauthUrl);
 *   }
 *   return;
 * }
 * if (isProofExpiringSoon(result.expiresAt)) {
 *   console.warn("Proof expiring in < 30 days — prompt for renewal");
 * }
 * ```
 */
async function verifyAgent(agentKey, options, rpcUrl) {
    if (!/^0x[0-9a-fA-F]{64}$/.test(agentKey)) {
        throw new TypeError(`agentKey must be a 0x-prefixed 32-byte hex string; received "${agentKey}"`);
    }
    // Resolve the RPC URL: use the provided override, or fall back to the
    // network whose registry address matches, or finally to Celo mainnet.
    const resolvedRpcUrl = rpcUrl ??
        Object.values(constants_1.NETWORKS).find((n) => n.registryAddress.toLowerCase() ===
            options.registryAddress.toLowerCase())?.rpcUrl ??
        constants_1.NETWORKS[constants_1.DEFAULT_NETWORK].rpcUrl;
    const provider = new ethers_1.ethers.JsonRpcProvider(resolvedRpcUrl);
    const registry = (0, contract_types_1.typedRegistry)(options.registryAddress, provider);
    // Step 1: resolve agent key → agentId
    const agentId = await registry.getAgentId(agentKey);
    if (agentId === 0n) {
        return { verified: false, reason: "NOT_REGISTERED" };
    }
    // Step 2: check whether the agent has a human proof at all
    const hasProof = await registry.hasHumanProof(agentId);
    if (!hasProof) {
        return { verified: false, reason: "NO_HUMAN_PROOF" };
    }
    // Step 3: check expiry (proofExpiresAt returns 0 if the proof never expires)
    const expiresAtSecs = await registry.proofExpiresAt(agentId);
    const nowSecs = BigInt(Math.floor(Date.now() / 1000));
    if (expiresAtSecs > 0n && nowSecs >= expiresAtSecs) {
        return {
            verified: false,
            reason: "PROOF_EXPIRED",
            expiredAt: new Date(Number(expiresAtSecs) * 1000),
            reauthUrl: buildReauthUrl(agentId, options),
        };
    }
    return {
        verified: true,
        agentId,
        expiresAt: expiresAtSecs > 0n ? new Date(Number(expiresAtSecs) * 1000) : null,
    };
}
