// @selfxyz/agent-sdk/dist/SelfAgent.js
"use strict";
// SPDX-FileCopyrightText: 2025-2026 Social Connect Labs, Inc.
// SPDX-License-Identifier: BUSL-1.1
// NOTE: Converts to Apache-2.0 on 2029-06-11 per LICENSE.
Object.defineProperty(exports, "__esModule", { value: true });
exports.SelfAgent = void 0;
const ethers_1 = require("ethers");
const constants_1 = require("./constants");
const agentCard_1 = require("./agentCard");
const signing_1 = require("./signing");
const registration_flow_1 = require("./registration-flow");
const contract_types_1 = require("./contract-types");
/**
 * Agent-side SDK for Self Agent ID.
 *
 * The agent's on-chain identity is its Ethereum address, zero-padded to bytes32:
 *   agentKey = zeroPadValue(address, 32)
 *
 * For off-chain authentication, the agent signs each request with its private key.
 * Services verify the signature, recover the signer address, and check on-chain status.
 *
 * Usage:
 * ```ts
 * // Mainnet (default)
 * const agent = new SelfAgent({ privateKey: "0x..." });
 *
 * // Testnet
 * const agent = new SelfAgent({ privateKey: "0x...", network: "testnet" });
 *
 * const registered = await agent.isRegistered();
 * const response = await agent.fetch("https://api.example.com/data");
 * ```
 */
class SelfAgent {
    constructor(config) {
        if (!config.privateKey && !config.signer) {
            throw new Error("Either privateKey or signer must be provided");
        }
        const net = constants_1.NETWORKS[config.network ?? constants_1.DEFAULT_NETWORK];
        const provider = new ethers_1.ethers.JsonRpcProvider(config.rpcUrl ?? net.rpcUrl);
        if (config.signer) {
            this.wallet = config.signer;
        }
        else {
            this.wallet = new ethers_1.ethers.Wallet(config.privateKey, provider);
        }
        this.registry = (0, contract_types_1.typedRegistry)(config.registryAddress ?? net.registryAddress, provider);
        // Agent key = address zero-padded to 32 bytes (matches on-chain derivation)
        this._agentKey = ethers_1.ethers.zeroPadValue(this.wallet.address, 32);
    }
    /** The agent's on-chain key (bytes32) — zero-padded address */
    get agentKey() {
        return this._agentKey;
    }
    /** The agent's Ethereum address */
    get address() {
        return this.wallet.address;
    }
    /** Check if this agent is registered and verified on-chain */
    async isRegistered() {
        return this.registry.isVerifiedAgent(this._agentKey);
    }
    /** Get full agent info from the registry */
    async getInfo() {
        const agentId = await this.registry.getAgentId(this._agentKey);
        if (agentId === 0n) {
            return {
                address: this.wallet.address,
                agentKey: this._agentKey,
                agentId: 0n,
                isVerified: false,
                nullifier: 0n,
                agentCount: 0n,
                proofExpiresAt: 0n,
                isProofFresh: false,
                daysUntilExpiry: -1,
                isExpiringSoon: false,
                siblingAgentIds: [],
            };
        }
        const [isVerified, nullifier] = await Promise.all([
            this.registry.hasHumanProof(agentId),
            this.registry.getHumanNullifier(agentId),
        ]);
        const [agentCount, proofExpiresAt, isProofFresh, siblingAgentIds] = await Promise.all([
            this.registry.getAgentCountForHuman(nullifier),
            this.registry.proofExpiresAt(agentId),
            this.registry.isProofFresh(agentId),
            this.registry.getAgentsForNullifier(nullifier),
        ]);
        const now = BigInt(Math.floor(Date.now() / 1000));
        const daysUntilExpiry = proofExpiresAt > 0n
            ? Number((proofExpiresAt - now) / 86400n)
            : -1;
        const isExpiringSoon = daysUntilExpiry >= 0 && daysUntilExpiry <= 30;
        return {
            address: this.wallet.address,
            agentKey: this._agentKey,
            agentId,
            isVerified,
            nullifier,
            agentCount,
            proofExpiresAt,
            isProofFresh,
            daysUntilExpiry,
            isExpiringSoon,
            siblingAgentIds,
        };
    }
    /**
     * Generate authentication headers for a request.
     *
     * The service recovers the signer address from the signature,
     * converts it to a bytes32 agent key, and checks on-chain status.
     *
     * Signature covers: keccak256(timestamp + method + canonicalPathAndQuery + bodyHash)
     */
    async signRequest(method, url, body) {
        const timestamp = Date.now().toString();
        const message = (0, signing_1.computeSigningMessage)(timestamp, method, url, body);
        const signature = await this.wallet.signMessage(ethers_1.ethers.getBytes(message));
        return {
            [constants_1.HEADERS.ADDRESS]: this.wallet.address,
            [constants_1.HEADERS.SIGNATURE]: signature,
            [constants_1.HEADERS.TIMESTAMP]: timestamp,
        };
    }
    /**
     * Wrapper around global fetch that automatically adds agent signature headers.
     */
    async fetch(url, options = {}) {
        const method = (options.method || "GET").toUpperCase();
        const body = typeof options.body === "string" ? options.body : undefined;
        const headers = await this.signRequest(method, url, body);
        return globalThis.fetch(url, {
            ...options,
            method,
            headers: {
                ...Object.fromEntries(Object.entries(options.headers || {})),
                ...headers,
            },
        });
    }
    // ─── A2A Agent Card Methods ──────────────────────────────────────────────
    /** Read the A2A Agent Card from on-chain metadata (if set) */
    async getAgentCard() {
        const agentId = await this.registry.getAgentId(this._agentKey);
        if (agentId === 0n)
            return undefined;
        const raw = await this.registry.getAgentMetadata(agentId);
        if (!raw)
            return undefined;
        try {
            const parsed = JSON.parse(raw);
            if (!parsed || typeof parsed !== "object")
                return undefined;
            if (("type" in parsed &&
                parsed.type ===
                    "https://eips.ethereum.org/EIPS/eip-8004#registration-v1") ||
                "a2aVersion" in parsed) {
                return parsed;
            }
        }
        catch (err) {
            console.warn("[SelfAgent] Failed to parse agent card metadata:", err);
        }
        return undefined;
    }
    /**
     * Build and write an A2A Agent Card to on-chain metadata.
     * Auto-populates selfProtocol fields from on-chain data.
     * Returns the transaction hash.
     */
    async setAgentCard(fields) {
        const agentId = await this.registry.getAgentId(this._agentKey);
        if (agentId === 0n)
            throw new Error("Agent not registered");
        const providerAddr = await this.registry.getProofProvider(agentId);
        if (!providerAddr || providerAddr === ethers_1.ethers.ZeroAddress) {
            throw new Error("Agent has no proof provider — cannot build card");
        }
        const provider = (0, contract_types_1.typedProvider)(providerAddr, this.registry.runner);
        const card = await (0, agentCard_1.buildAgentCard)(Number(agentId), this.registry, provider, fields);
        const registryWithSigner = this.registry.connect(this.wallet);
        const tx = (await registryWithSigner.updateAgentMetadata(agentId, JSON.stringify(card)));
        await tx.wait();
        return tx.hash;
    }
    /** Returns a data: URI containing the base64-encoded Agent Card JSON */
    async toAgentCardDataURI() {
        const card = await this.getAgentCard();
        if (!card)
            throw new Error("No A2A Agent Card set");
        const json = JSON.stringify(card);
        const encoded = btoa(json);
        return `data:application/json;base64,${encoded}`;
    }
    /** Read ZK-attested credentials for this agent from on-chain */
    async getCredentials() {
        const agentId = await this.registry.getAgentId(this._agentKey);
        if (agentId === 0n)
            return undefined;
        try {
            return (await this.registry.getAgentCredentials(agentId));
        }
        catch {
            return undefined;
        }
    }
    /** Read the verification strength score from the provider that verified this agent */
    async getVerificationStrength() {
        const agentId = await this.registry.getAgentId(this._agentKey);
        if (agentId === 0n)
            return 0;
        const providerAddr = await this.registry.getProofProvider(agentId);
        if (providerAddr === ethers_1.ethers.ZeroAddress)
            return 0;
        const provider = (0, contract_types_1.typedProvider)(providerAddr, this.registry.runner);
        const strength = await provider.verificationStrength();
        return Number(strength);
    }
    // ─── Registration & Deregistration (REST API) ─────────────────────────────
    /**
     * Request agent registration through the Self Agent ID REST API.
     *
     * Returns a session with a QR code / deep link for the human to scan,
     * plus a `waitForCompletion()` method that polls until on-chain verification.
     *
     * ```ts
     * const session = await SelfAgent.requestRegistration({
     *   mode: "linked",
     *   network: "mainnet",
     *   disclosures: { minimumAge: 18, ofac: true },
     *   humanAddress: "0x...",
     * });
     *
     * console.log(session.deepLink);       // open in Self app
     * console.log(session.humanInstructions); // tell human what to do
     *
     * const result = await session.waitForCompletion();
     * const agent = new SelfAgent({ privateKey: await session.exportKey() });
     * ```
     */
    static async requestRegistration(opts) {
        return (0, registration_flow_1.requestRegistration)(opts);
    }
    /**
     * Get agent info from the public REST API (no private key needed).
     */
    static async getAgentInfo(agentId, opts) {
        return (0, registration_flow_1.getAgentInfo)(agentId, opts);
    }
    /**
     * Get all agents registered for a human address from the public REST API.
     */
    static async getAgentsForHuman(address, opts) {
        return (0, registration_flow_1.getAgentsForHuman)(address, opts);
    }
    /**
     * Request deregistration for this agent through the Self Agent ID REST API.
     *
     * Returns a session with a QR code / deep link for the human to scan,
     * plus a `waitForCompletion()` method that polls until on-chain removal.
     *
     * ```ts
     * const session = await agent.requestDeregistration();
     * console.log(session.deepLink);
     * await session.waitForCompletion();
     * ```
     */
    async requestDeregistration(opts) {
        const network = this.networkName();
        return (0, registration_flow_1.requestDeregistration)({
            network,
            agentAddress: this.wallet.address,
            apiBase: opts?.apiBase,
        });
    }
    /**
     * Fetch this agent's credentials from the public REST API.
     * Returns null if the agent is not registered or has no credentials.
     */
    async getCredentialsFromApi(opts) {
        try {
            const info = await this.getInfo();
            if (info.agentId === 0n)
                return null;
            const network = this.networkName();
            const apiInfo = await (0, registration_flow_1.getAgentInfo)(Number(info.agentId), {
                network,
                apiBase: opts?.apiBase,
            });
            return apiInfo.credentials ?? null;
        }
        catch {
            return null;
        }
    }
    /** Resolve the network name from the registry address */
    networkName() {
        const registryAddr = this.registry.target.toLowerCase();
        for (const [name, config] of Object.entries(constants_1.NETWORKS)) {
            if (config.registryAddress.toLowerCase() === registryAddr) {
                return name;
            }
        }
        return constants_1.DEFAULT_NETWORK;
    }
}
exports.SelfAgent = SelfAgent;
