// @selfxyz/agent-sdk/dist/a2a/client.js
"use strict";
// SPDX-FileCopyrightText: 2025-2026 Social Connect Labs, Inc.
// SPDX-License-Identifier: BUSL-1.1
// NOTE: Converts to Apache-2.0 on 2029-06-11 per LICENSE.
Object.defineProperty(exports, "__esModule", { value: true });
exports.A2AClient = exports.A2AError = void 0;
const types_1 = require("./types");
/** Error thrown when an A2A JSON-RPC response contains an error. */
class A2AError extends Error {
    constructor(code, message, data) {
        super(message);
        this.name = "A2AError";
        this.code = code;
        this.data = data;
    }
}
exports.A2AError = A2AError;
/**
 * A2A protocol client. Sends JSON-RPC 2.0 requests to an A2A agent endpoint.
 */
class A2AClient {
    constructor(options) {
        this.nextId = 1;
        if (!options.agentUrl && !options.agentCard?.url) {
            throw new Error("A2AClient requires either agentUrl or an agentCard with a url");
        }
        this.agentUrl = options.agentUrl ?? options.agentCard.url;
        this.cachedCard = options.agentCard;
        this.headers = options.headers ?? {};
    }
    /**
     * Fetch and cache the A2A agent card from the well-known endpoint.
     * If a baseUrl is provided, fetches from `{baseUrl}/.well-known/agent-card.json`.
     * Otherwise uses the configured agentUrl's origin.
     */
    async fetchAgentCard(baseUrl) {
        const origin = baseUrl ?? new URL(this.agentUrl).origin;
        const cardUrl = `${origin.replace(/\/$/, "")}/.well-known/agent-card.json`;
        const res = await fetch(cardUrl, {
            headers: { Accept: "application/json", ...this.headers },
        });
        if (!res.ok) {
            throw new A2AError(types_1.A2AErrorCodes.INVALID_AGENT_CARD, `Failed to fetch agent card: ${res.status} ${res.statusText}`);
        }
        this.cachedCard = (await res.json());
        return this.cachedCard;
    }
    /** Returns the cached agent card, or undefined if not yet fetched. */
    getAgentCard() {
        return this.cachedCard;
    }
    /** Send a message to the agent. Returns the resulting Task. */
    async sendMessage(message, metadata) {
        return this.rpc("message/send", {
            message,
            metadata,
        });
    }
    /** Get the current state of a task. */
    async getTask(taskId, historyLength, metadata) {
        return this.rpc("tasks/get", {
            id: taskId,
            historyLength,
            metadata,
        });
    }
    /** Cancel a task. */
    async cancelTask(taskId, metadata) {
        return this.rpc("tasks/cancel", {
            id: taskId,
            metadata,
        });
    }
    /**
     * Send a message and receive streaming SSE events.
     * Returns an async iterable of A2AStreamEvent objects.
     */
    async *sendMessageStream(message, metadata) {
        const body = JSON.stringify({
            jsonrpc: "2.0",
            id: this.nextId++,
            method: "message/stream",
            params: { message, metadata },
        });
        const res = await fetch(this.agentUrl, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
                Accept: "text/event-stream",
                ...this.headers,
            },
            body,
        });
        if (!res.ok) {
            throw new A2AError(types_1.A2AErrorCodes.INTERNAL_ERROR, `Stream request failed: ${res.status} ${res.statusText}`);
        }
        if (!res.body) {
            throw new A2AError(types_1.A2AErrorCodes.UNSUPPORTED_OPERATION, "Response body is not a readable stream");
        }
        // Parse SSE from the response body
        const reader = res.body.getReader();
        const decoder = new TextDecoder();
        let buffer = "";
        try {
            while (true) {
                const { done, value } = await reader.read();
                if (done)
                    break;
                buffer += decoder.decode(value, { stream: true });
                const lines = buffer.split("\n");
                // Keep the last incomplete line in the buffer
                buffer = lines.pop() ?? "";
                let eventData = "";
                for (const line of lines) {
                    if (line.startsWith("data: ")) {
                        eventData += line.slice(6);
                    }
                    else if (line === "" && eventData) {
                        // Empty line = end of event
                        yield JSON.parse(eventData);
                        eventData = "";
                    }
                }
            }
        }
        finally {
            reader.releaseLock();
        }
    }
    // ─── Private ────────────────────────────────────────────────────────────
    async rpc(method, params) {
        const body = JSON.stringify({
            jsonrpc: "2.0",
            id: this.nextId++,
            method,
            params,
        });
        const res = await fetch(this.agentUrl, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
                Accept: "application/json",
                ...this.headers,
            },
            body,
        });
        if (!res.ok) {
            throw new A2AError(types_1.A2AErrorCodes.INTERNAL_ERROR, `HTTP ${res.status} ${res.statusText}`);
        }
        const json = (await res.json());
        if (json.error) {
            throw new A2AError(json.error.code, json.error.message, json.error.data);
        }
        return json.result;
    }
}
exports.A2AClient = A2AClient;
