// @selfxyz/agent-sdk/dist/a2a/server.js
"use strict";
// SPDX-FileCopyrightText: 2025-2026 Social Connect Labs, Inc.
// SPDX-License-Identifier: BUSL-1.1
// NOTE: Converts to Apache-2.0 on 2029-06-11 per LICENSE.
Object.defineProperty(exports, "__esModule", { value: true });
exports.A2AServer = void 0;
const types_1 = require("./types");
// ─── A2A Server ─────────────────────────────────────────────────────────────
/**
 * Framework-agnostic A2A JSON-RPC 2.0 server.
 *
 * Takes raw request objects and returns response objects.
 * Plug into any HTTP framework (Express, Fastify, Next.js, etc.)
 * by parsing the request body and passing it to `handleRequest()`.
 */
class A2AServer {
    constructor(handler) {
        this.tasks = new Map();
        this.handler = handler;
    }
    /** Get a stored task by ID (from the in-memory store). */
    getStoredTask(taskId) {
        return this.tasks.get(taskId);
    }
    /**
     * Process a raw JSON-RPC request object and return a JSON-RPC response.
     * This is the main entry point — call this from your HTTP handler.
     */
    async handleRequest(request) {
        // Validate basic JSON-RPC structure
        if (!isObject(request)) {
            return errorResponse(0, types_1.A2AErrorCodes.PARSE_ERROR, "Parse error");
        }
        const req = request;
        if (req.jsonrpc !== "2.0") {
            return errorResponse(toId(req.id), types_1.A2AErrorCodes.INVALID_REQUEST, "Invalid JSON-RPC version");
        }
        if (typeof req.method !== "string") {
            return errorResponse(toId(req.id), types_1.A2AErrorCodes.INVALID_REQUEST, "Missing or invalid method");
        }
        const id = toId(req.id);
        try {
            switch (req.method) {
                case "message/send":
                    return await this.handleSendMessage(id, req.params);
                case "tasks/get":
                    return await this.handleGetTask(id, req.params);
                case "tasks/cancel":
                    return await this.handleCancelTask(id, req.params);
                case "message/stream":
                    // Streaming is not handled via simple request/response.
                    // Frameworks should detect this method and handle SSE separately.
                    return errorResponse(id, types_1.A2AErrorCodes.UNSUPPORTED_OPERATION, "Streaming must be handled via SSE. Use handleStreamRequest() or implement SSE in your framework.");
                default:
                    return errorResponse(id, types_1.A2AErrorCodes.METHOD_NOT_FOUND, `Unknown method: ${req.method}`);
            }
        }
        catch (err) {
            return errorResponse(id, types_1.A2AErrorCodes.INTERNAL_ERROR, err instanceof Error ? err.message : "Internal error");
        }
    }
    // ─── Method Handlers ──────────────────────────────────────────────────
    async handleSendMessage(id, params) {
        if (!isObject(params)) {
            return errorResponse(id, types_1.A2AErrorCodes.INVALID_PARAMS, "Missing params");
        }
        const p = params;
        if (!p.message || !Array.isArray(p.message.parts)) {
            return errorResponse(id, types_1.A2AErrorCodes.INVALID_PARAMS, "Invalid message: must include parts array");
        }
        if (!["user", "agent"].includes(p.message.role)) {
            return errorResponse(id, types_1.A2AErrorCodes.INVALID_PARAMS, 'Invalid message role: must be "user" or "agent"');
        }
        const task = await this.handler.onMessage(p.message, p.metadata);
        this.tasks.set(task.id, task);
        return {
            jsonrpc: "2.0",
            id,
            result: task,
        };
    }
    async handleGetTask(id, params) {
        if (!isObject(params)) {
            return errorResponse(id, types_1.A2AErrorCodes.INVALID_PARAMS, "Missing params");
        }
        const p = params;
        if (typeof p.id !== "string") {
            return errorResponse(id, types_1.A2AErrorCodes.INVALID_PARAMS, "Missing or invalid task id");
        }
        try {
            const task = await this.handler.onGetTask(p.id, p.historyLength);
            this.tasks.set(task.id, task);
            return { jsonrpc: "2.0", id, result: task };
        }
        catch (err) {
            if (err instanceof Error &&
                err.message.toLowerCase().includes("not found")) {
                return errorResponse(id, types_1.A2AErrorCodes.TASK_NOT_FOUND, err.message);
            }
            throw err;
        }
    }
    async handleCancelTask(id, params) {
        if (!isObject(params)) {
            return errorResponse(id, types_1.A2AErrorCodes.INVALID_PARAMS, "Missing params");
        }
        const p = params;
        if (typeof p.id !== "string") {
            return errorResponse(id, types_1.A2AErrorCodes.INVALID_PARAMS, "Missing or invalid task id");
        }
        try {
            const task = await this.handler.onCancelTask(p.id);
            this.tasks.set(task.id, task);
            return { jsonrpc: "2.0", id, result: task };
        }
        catch (err) {
            if (err instanceof Error &&
                err.message.toLowerCase().includes("not found")) {
                return errorResponse(id, types_1.A2AErrorCodes.TASK_NOT_FOUND, err.message);
            }
            if (err instanceof Error &&
                err.message.toLowerCase().includes("not cancelable")) {
                return errorResponse(id, types_1.A2AErrorCodes.TASK_NOT_CANCELABLE, err.message);
            }
            throw err;
        }
    }
}
exports.A2AServer = A2AServer;
// ─── Helpers ────────────────────────────────────────────────────────────────
function isObject(val) {
    return typeof val === "object" && val !== null && !Array.isArray(val);
}
function toId(val) {
    if (typeof val === "string" || typeof val === "number")
        return val;
    return 0;
}
function errorResponse(id, code, message, data) {
    return {
        jsonrpc: "2.0",
        id,
        error: { code, message, ...(data !== undefined ? { data } : {}) },
    };
}
